File: architecture-skills/object-oriented-programming-paradigm/examples/coupling-and-cohesion/coupling-and-cohesion.php
<?php
// Coupling and Cohesion in PHP.
//
// A god OrderProcessor that computes tax, validates, and ships couples
// three volatile concerns. Focused classes with high cohesion let a tax
// change touch only OrderCalculator.

declare(strict_types=1);

// --- Before (tightly coupled, sketched): ---
//
// final class OrderProcessor {
//     public function process(Order $order): float {
//         $total = $order->subtotal * 1.20;  // tax
//         if ($total <= 0) throw ...;        // validation
//         echo "shipped {$order->id}";       // shipping
//         return $total;
//     }
// }

final class Order
{
    public function __construct(
        public readonly string $id,
        public readonly float $subtotal,
    ) {}
}

/** Finance's class: only tax-rule changes touch this file. */
final class OrderCalculator
{
    public function __construct(private readonly float $taxRate) {}

    public function total(Order $order): float
    {
        return $order->subtotal * (1 + $this->taxRate);
    }
}

/** Compliance's class: only validation rules change touch this file. */
final class OrderValidator
{
    public function validate(Order $order, float $total): void
    {
        if ($order->subtotal <= 0) {
            throw new InvalidArgumentException('subtotal must be positive');
        }
        if ($total <= 0) {
            throw new InvalidArgumentException('total must be positive');
        }
    }
}

/** Fulfillment's class: only carrier/API changes touch this file. */
final class OrderShipper
{
    public function ship(Order $order): string
    {
        return "shipped {$order->id}";
    }
}

/** Thin orchestrator: wires cohesive pieces, owns no domain rules. */
final class OrderService
{
    public function __construct(
        private readonly OrderCalculator $calculator,
        private readonly OrderValidator $validator,
        private readonly OrderShipper $shipper,
    ) {}

    public function process(Order $order): float
    {
        $total = $this->calculator->total($order);
        $this->validator->validate($order, $total);
        echo $this->shipper->ship($order) . PHP_EOL;
        return $total;
    }
}

$order = new Order('ORD-204', 80.00);

$service = new OrderService(
    new OrderCalculator(0.20),
    new OrderValidator(),
    new OrderShipper(),
);
$total = $service->process($order);
printf("total with 20%% tax: $%.2f\n", $total);

$newRateService = new OrderService(
    new OrderCalculator(0.08),
    new OrderValidator(),
    new OrderShipper(),
);
$newTotal = $newRateService->process($order);
printf("total with 8%% tax:  $%.2f  (only OrderCalculator changed)\n", $newTotal);

File: architecture-skills/object-oriented-programming-paradigm/examples/coupling-and-cohesion/coupling-and-cohesion.cpp
// Coupling and Cohesion in C++.
//
// A god OrderProcessor that computes tax, validates, and ships couples
// three volatile concerns. Focused classes with high cohesion let a tax
// change touch only OrderCalculator.

#include <iostream>
#include <stdexcept>
#include <string>

// --- Before (tightly coupled, sketched): ---
//
// class OrderProcessor {
// public:
//     double process(Order& o) {
//         double total = o.subtotal * 1.20;  // tax
//         if (total <= 0) throw ...;         // validation
//         std::cout << "shipped " << o.id;   // shipping
//         return total;
//     }
// };

struct Order {
    std::string id;
    double subtotal;
};

// Finance's class: only tax-rule changes touch this file.
class OrderCalculator {
public:
    explicit OrderCalculator(double taxRate) : taxRate_(taxRate) {}

    double total(const Order& order) const {
        return order.subtotal * (1.0 + taxRate_);
    }

private:
    double taxRate_;
};

// Compliance's class: only validation rules change touch this file.
class OrderValidator {
public:
    void validate(const Order& order, double total) const {
        if (order.subtotal <= 0) throw std::invalid_argument("subtotal must be positive");
        if (total <= 0) throw std::invalid_argument("total must be positive");
    }
};

// Fulfillment's class: only carrier/API changes touch this file.
class OrderShipper {
public:
    std::string ship(const Order& order) const {
        return "shipped " + order.id;
    }
};

// Thin orchestrator: wires cohesive pieces together.
class OrderService {
public:
    OrderService(OrderCalculator calc, OrderValidator validator, OrderShipper shipper)
        : calc_(std::move(calc)), validator_(std::move(validator)), shipper_(std::move(shipper)) {}

    double process(const Order& order) const {
        double total = calc_.total(order);
        validator_.validate(order, total);
        std::cout << shipper_.ship(order) << "\n";
        return total;
    }

private:
    OrderCalculator calc_;
    OrderValidator validator_;
    OrderShipper shipper_;
};

int main() {
    Order order{"ORD-204", 80.00};

    OrderService service(OrderCalculator(0.20), OrderValidator{}, OrderShipper{});
    double total = service.process(order);
    std::cout << "total with 20% tax: $" << total << "\n";

    OrderService newRateService(OrderCalculator(0.08), OrderValidator{}, OrderShipper{});
    double newTotal = newRateService.process(order);
    std::cout << "total with 8% tax:  $" << newTotal
              << "  (only OrderCalculator changed)\n";
    return 0;
}
